### PR TITLE
Initialization phase for a Subgraph

### PR DESCRIPTION
**Pull request by jona**

### CODE REVIEW

File: graph.config.yml [hunk 1]
```diff
@@ -48,6 +48,8 @@ contracts:
       reserveRatio: 500000 # 50% (parts per million)
       curationTaxPercentage: 25000 # 2.5% (parts per million)
       minimumCurationDeposit: "1000000000000000000" # 1 GRT
+      initializationPeriod: 432000 # Seconds = 5 Days
+      initializationExitPeriod: 2592000 # Seconds = 30 Days
```
Comment (jona) [inline graph.config.yml:52]: These will need to be adjusted once we finalize the GIP. They will also be affected by the upcoming Capital Gains Tax GIP as they both work in harmony to balance out incentives/disincentives.

File: contracts/curation/Curation.sol [hunk 2]
```diff
@@ -463,4 +530,28 @@ contract Curation is CurationV1Storage, GraphUpgradeable, ICuration {
...
+     */
+    function _getEffectiveReserveRatio(uint256 _createdAt) private view returns (uint32) {
+        uint32 effectiveReserveRatio = defaultReserveRatio;
+
+        if (block.timestamp <= _createdAt.add(initializationPeriod)) {
+            effectiveReserveRatio = MAX_PPM;
+        } else if (
+            block.timestamp <= (_createdAt.add(initializationPeriod).add(initializationExitPeriod))
+        ) {
+            uint256 blockDiff = block.timestamp.sub(_createdAt.add(initializationPeriod));
```
Comment (abarmat) [inline contracts/curation/Curation.sol:547]: Just a warning, this is a `timeDiff` now instead of blocks

File: contracts/curation/Curation.sol [hunk 3]
```diff
@@ -395,6 +456,8 @@ contract Curation is CurationV1Storage, GraphUpgradeable, ICuration {
         // Get curation pool tokens and signal
         CurationPool memory curationPool = pools[_subgraphDeploymentID];
 
+        uint32 _effectiveReserveRatio = _getEffectiveReserveRatio(curationPool.createdAt);
```
Comment (abarmat) [inline contracts/curation/Curation.sol:459]: Something that happens with this implementation is that it breaks the possibility to mint at the Curation contract level. 

If the Curation.mint() function is used directly by any user the `createdAt` variable will be zero. I think we should store `createdAt` on the subgraph struct whenever the first mint happens as part of initialization.

File: contracts/curation/Curation.sol [hunk 4]
```diff
@@ -463,4 +530,28 @@ contract Curation is CurationV1Storage, GraphUpgradeable, ICuration {
...
+
+    /**
+     * @dev Calculate reserve ratio based on initialization phase
+     * @param _createdAt When NameCurationPool was created
+     * @return Reserve ratio
+     */
+    function _getEffectiveReserveRatio(uint256 _createdAt) private view returns (uint32) {
+        uint32 effectiveReserveRatio = defaultReserveRatio;
+
+        if (block.timestamp <= _createdAt.add(initializationPeriod)) {
```
Comment (abarmat) [inline contracts/curation/Curation.sol:542]: We could store this intermediate calculation in a variable as it is reused in different places `_createdAt.add(initializationPeriod)`

File: contracts/curation/Curation.sol [hunk 5]
```diff
@@ -463,4 +530,28 @@ contract Curation is CurationV1Storage, GraphUpgradeable, ICuration {
...
         }
     }
+
+    /**
+     * @dev Calculate reserve ratio based on initialization phase
+     * @param _createdAt When NameCurationPool was created
+     * @return Reserve ratio
+     */
+    function _getEffectiveReserveRatio(uint256 _createdAt) private view returns (uint32) {
+        uint32 effectiveReserveRatio = defaultReserveRatio;
```
Comment (abarmat) [inline contracts/curation/Curation.sol:540]: One issue I see with reading directly from `defaultReserveRatio` is that if Governance changes the reserve ratio at any point all the calculations from previously initialized subgraphs will change. 

The current behaviour is to store the reserve ratio to use in `Subgraph.reserveRatio` whenever it is initialized. I believe we should still do that during the first mint and use it from here.

What do you think?

File: contracts/discovery/GNS.sol [hunk 6]
```diff
@@ -286,6 +289,10 @@ contract GNS is GNSV1Storage, GraphUpgradeable, IGNS, Multicall {
             "GNS: Cannot publish a new version with the same subgraph deployment ID"
         );
 
+        NameCurationPool storage namePool = nameSignals[_graphAccount][_subgraphNumber];
+
+        curation().setCreatedAt(_subgraphDeploymentID, namePool.createdAt);
```
Comment (abarmat) [inline contracts/discovery/GNS.sol:294]: I don't feel comfortable with this GNS->Curation interaction that makes Curation completely dependent on the GNS, but I'm trying to think if there is a better option and couldn't think of anything better yet.

File: contracts/curation/Curation.sol (hunk 4 above)
Comment (abarmat) [reply contracts/curation/Curation.sol:542]: When we upgrade the Curation contract with this new implementation, all the existing subgraphs deployments will have createdAt set at 0. Maybe we should add if-clause to return the plain reserve ratio when Subgraph.createdAt = 0

**Pull request closed without merging**